### PR TITLE
streamingccl: physical replication stream from given timestamp 

### PR DESCRIPTION
**Pull request by stevendanna**
This adds a RESUME TIMESTAMP option to CREATE VIRTUAL CLUSTER FROM
REPLICATION.  When provided, we allow the user to start a replication
stream into an _existing_ virtual cluster.

When the resume timestamp is provided, the replication stream will be
started from that timestamp, with no initial scan. To facilitate this,
we add a new argument to crdb_internal.start_replication_stream that
allows us to pass a destination-choosen start timestamp.

To avoid various catastrophic mistakes, we only allow this when:

- The destination tenant must be in service mode None

- The provided resume timestamp equals the last recorded "revert
  timestamp" of the destination tenant. The revert timestamp is set when
  the tenant has been forcibly reverted to a particular timestamp and is
  cleared when the tenant is modified in a way that may invalidate a
  resumption from that timestamp.

- If the source tenant has a PreviousSourceTenant set, the new
  destination must match that previous source tenant field.

WARNING: Using this correctly requires that the stream is resumed
before garbage collection has progressed past the given resume
timestamp. During normal operation, the replication stream maintains a
protected timestamp to ensure this is the case. However, when
resuming using this new feature, we have no such guarantee.

Epic: none

Release note: None

### CODE REVIEW

**Comment by stevendanna** [comment]
I need to write some more tests and clean up a bit, but wanted to get this out to get review on the overall shape of it.

File: pkg/sql/parser/testdata/create_virtual_cluster [hunk 1]
```diff
@@ -94,6 +94,14 @@ CREATE VIRTUAL CLUSTER ("destination-hyphen") FROM REPLICATION OF ("source-hyphe
 CREATE VIRTUAL CLUSTER "destination-hyphen" FROM REPLICATION OF "source-hyphen" ON '_' WITH RETENTION = '_' -- literals removed
 CREATE VIRTUAL CLUSTER _ FROM REPLICATION OF _ ON 'pgurl' WITH RETENTION = '36h' -- identifiers removed
 
+parse
```
Comment (adityamaru) [inline pkg/sql/parser/testdata/create_virtual_cluster:97]: WITH RESUME here?

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_planning.go [hunk 2]
```diff
@@ -158,58 +170,124 @@ func ingestionPlanHook(
...
+				)
+			}
+			if tenantRecord.LastRevertTenantTimestamp.IsEmpty() {
+				return errors.Newf("cannot start replication for tenant %q (%d) with no last revert timestamp found; likely that this tenant cannot be safely streamed into",
+					tenantRecord.Name,
+					tenantRecord.ID,
+				)
+			}
+			if !tenantRecord.LastRevertTenantTimestamp.Equal(options.resumeTimestamp) {
+				return errors.Newf("cannot start replication for tenant %q (%d) with resume timestamp %s that doesn't match last resume timestamp %s",
```
Comment (adityamaru) [inline pkg/ccl/streamingccl/streamingest/stream_ingestion_planning.go:246]: nit: that doesn't match last s/resume/revert/g timestamp?

File: pkg/repstream/streampb/stream.proto [hunk 3]
```diff
@@ -35,6 +35,46 @@ message ReplicationProducerSpec {
...
+  // ReplicationStartTime is the initial timestamp from which the replication
+  // producer job will begin streaming MVCC revisions. This timestamp is picked
+  // once when the replication producer job is created, and is never updated
+  // through the lifetime of a replication stream. This will be the timestamp as
+  // of which each partition will perform its initial rangefeed scan.
+  util.hlc.Timestamp replication_start_time = 1 [(gogoproto.nullable) = false];
+
+
+  // ClusterID identifies the /requesting/ cluster. If set, this is
+  // compred to the ClusterID in the tenant record's
```
Comment (adityamaru) [inline pkg/repstream/streampb/stream.proto:63]: s/compred/compared/g

File: pkg/ccl/streamingccl/streamproducer/stream_lifetime.go [hunk 4]
```diff
@@ -69,7 +74,36 @@ func startReplicationProducerJob(
...
+
+			tid := tenantRecord.PreviousSourceTenant.TenantID
+			if !req.TenantID.Equal(roachpb.TenantID{}) && !tid.Equal(roachpb.TenantID{}) {
+				if !req.TenantID.Equal(tid) {
+					return streampb.ReplicationProducerSpec{}, errors.Errorf("requesting tenant ID %s does not match previous source tenant ID %s",
+						req.TenantID, tid)
+				}
+			}
+		}
+		replicationStartTime = req.ReplicationStartTime
```
Comment (adityamaru) [inline pkg/ccl/streamingccl/streamproducer/stream_lifetime.go:101]: if tenantRecord.PreviousSourceTenant == nil when req.ReplicationStartTime is not empty, we should `errors.AssertionFailed` ?

File: pkg/ccl/streamingccl/streamingest/alter_replication_job.go [hunk 5]
```diff
@@ -131,15 +148,19 @@ func alterReplicationJobHook(
 			"only the system tenant can alter tenant")
 	}
 
+	if alterTenantStmt.Options.ResumeTimestamp != nil {
+		return nil, nil, nil, false, pgerror.New(pgcode.InvalidParameterValue, "resume timetamp cannot be altered")
```
Comment (adityamaru) [inline pkg/ccl/streamingccl/streamingest/alter_replication_job.go:152]: s/timetamp/timestamp/g

**Review by adityamaru — APPROVED** [review]

**Review by adityamaru — APPROVED** [review]
I don't have any blocking comments, this is really great. A roachtest or two for failback as future changes would be great but definitely a follow up.

**Review by adityamaru — APPROVED** [review]
I don't have any blocking comments, this is really great. A roachtest or two for failback as future changes would be great but definitely a follow up.

**Comment by stevendanna** [comment]
bors r=adityamaru

**Pull request merged**